Information/CenterOfMass.FCMacro

```diff
@@ -1,891 +1,966 @@
-        defcolor = QtGui.QColor(255, 255, 255)
-        self.spin.setStyleSheet('QDoubleSpinBox {background-color: %s}'%defcolor.name())
-
+        qty = self.parent.material_base.get(materialName, 0)
+        if qty:
+            materialDensity = Units.Quantity(qty).getValueAs(self.parent.unitForD)
+            self.spinDens.blockSignals(True)
+            # don't trigger on_spinDensity_changed
+            self.spinDens.setValue(materialDensity)
+            self.spinDens.blockSignals(False)
+            self.parent.computeCenterOfMass()
+            self.update_spinMass_only()
 
-def convert_density(density_siunits, unit):
-    """Convert the density in kg/m³ into another unit"""
-    tu = FreeCAD.Units.parseQuantity
-    return float(tu('{} kg/m^3'.format(density_siunits)) / tu(unit))
 
-
-# This is the widget which does almost all of the work
-# Widgets don't have close boxes, so closing is dealt with in
-# CenterofmassWindow.
 class CenterofmassWidget(QtGui.QWidget):
-    def __init__(self, parent=None):
-        super(CenterofmassWidget, self).__init__(parent)
-        self.parent = parent
+    """This is the widget which does almost all of the work.
+    Widgets don't have close boxes, so closing is dealt with in
+    CenterofmassWindow."""
+    def __init__(self, parent):
+        super().__init__(parent)
+        self.setObjectName(__Name__)
+        parent.setWindowTitle(__Name__ + ' ' + __Version__)
+        parent.setFont(g_font)
+        self.doc = FreeCAD.ActiveDocument
+        self.material_base = {}
         self.solid_count = 0
-        self.TotalCdG = [0, 0, 0]
-
-    tu = FreeCAD.Units.parseQuantity
-
-    material_database = {}
-
-    # Lay out the interactive elements
-    def initUI(self, parent):
-        global g_win_width
+        self.startup_failed = False
         global sel
-
-        self.massTot = 0.0
-        self.centerOfMass = [0.0, 0.0, 0.0]
-
-        font = QtGui.QFont()
-        font.setFamily('Times New Roman')
-        font.setPointSize(10)
-        font.setWeight(10)
-        font.setBold(True)
-
-        sel = FreeCADGui.Selection.getSelection()
-        objs  = []
-        self.solid_count = 0
-        for s in range(len(sel)):
-            if hasattr(sel[s], 'Shape'):
-                if sel[s].Shape.Volume == 0:
-                    error_dialog(u'invalid selection (select a solid or a mesh)')
-                    self.destroy()
-                else:
-                    self.solid_count += 1
-                    objs.append(sel[s].Shape)
-            elif hasattr(sel[s], 'Mesh'):
-                if sel[s].Mesh.Volume == 0:
-                    error_dialog(u'invalid selection (select a solid or a mesh)')
-                    self.destroy()
-                else:
-                    self.solid_count += 1
-                    objs.append(sel[s].Mesh)
+        sel = []
+        objs = self.validSelection()
+        while not self.solid_count and not self.startup_failed:
```

Comment:
`solid_count` is not a `bool`. Compare with a number to get a `bool`, i.e. `while self.solid_count > 0 and not self.startup_failed`.